### PR TITLE
ovl_En_Ik

### PR DESCRIPTION
**Pull request by Sirius902**
First miniboss! The reason there is a large static array inside a function is because otherwise I'd have to create useless copies of the vectors and colors below it, because apparently IDO puts static data in functions after non-static data.

### CODE REVIEW

**Review by Kenix3 — APPROVED** [review]
Looks good to me.

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 1]
```diff
@@ -1,15 +1,1401 @@
...
+        sp2C.x = this->actor.posRot.pos.x + Math_Sins(this->actor.shape.rot.y + 0x6A4) * 70.0f;
+        sp2C.z = this->actor.posRot.pos.z + Math_Coss(this->actor.shape.rot.y + 0x6A4) * 70.0f;
+        sp2C.y = this->actor.posRot.pos.y;
+        Audio_PlayActorSound2(&this->actor, NA_SE_EN_IRONNACK_HIT_GND);
+        func_8005AA1C(&globalCtx->mainCamera, 2, 0x19, 5);
+        func_800AA000(this->actor.xzDistFromLink, 0xFF, 0x14, 0x96);
+        func_80062CD4(globalCtx, &sp2C);
+    }
+
+    if ((17.0f < this->skelAnime.animCurrentFrame) && (this->skelAnime.animCurrentFrame < 23.0f)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:385]: imo variable > const would be better here, though i know some people are used to the other notation from math related stuff

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 2]
```diff
@@ -1,15 +1,1401 @@
...
+    if (SkelAnime_FrameUpdateMatrix(&this->skelAnime)) {
+        func_80A7506C(this);
+    }
+}
+
+void func_80A7506C(EnIk* this) {
+    f32 frames = SkelAnime_GetFrameCount(&D_060029FC.genericHeader);
+
+    this->unk_2FE = 0;
+    this->unk_2F9 = (s8)frames;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:404]: does s8 temp avoid a cast?

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 3]
```diff
@@ -1,15 +1,1401 @@
...
+    this->unk_2FC = 0;
+    EnIk_SetupAction(this, func_80A75260);
+}
+
+void func_80A75260(EnIk* this, GlobalContext* globalCtx) {
+    f32 temp_f0;
+
+    this->unk_300 += 0x1C2;
+    temp_f0 = Math_Sins(this->unk_300);
+    this->skelAnime.animPlaybackSpeed = ABS(temp_f0);
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:443]: temp needed?

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 4]
```diff
@@ -1,15 +1,1401 @@
...
+
+    if (this->skelAnime.animCurrentFrame > 11.0f) {
+        this->unk_2FF = 3;
+    }
+    if (((1.0f < this->skelAnime.animCurrentFrame) && (this->skelAnime.animCurrentFrame < 9.0f)) ||
+        ((13.0f < this->skelAnime.animCurrentFrame) && (this->skelAnime.animCurrentFrame < 18.0f))) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:449]: same here imo, but we dont really have a set convention on this for now so you dont have to

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 5]
```diff
@@ -1,15 +1,1401 @@
...
+        return;
+    }
+    Math_SmoothScaleMaxMinS(&this->actor.posRot.rot.y, this->actor.yawTowardsLink, 1, 0x7D0, 0);
+    if ((this->actor.params == 0) && (Math_Rand_ZeroOne() < 0.5f)) {
+        if (ABS((s16)(this->actor.yawTowardsLink - this->actor.shape.rot.y)) > 0x4000) {
+            func_80A754A0(this);
+        }
+    }
+    if ((this->actor.params != 0) && (this->unk_2FB != 0)) {
+        if ((prevHealth >= 11) && (this->actor.colChkInfo.health < 11)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:680]: think `> 10 && <= 10` would work

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 6]
```diff
@@ -1,15 +1,1401 @@
...
+
+void func_80A75FA0(Actor* thisx, GlobalContext* globalCtx) {
+    EnIk* this = THIS;
+    s32 pad;
+    Player* player = PLAYER;
+    u8 prevInvincibilityTimer;
+
+    this->unk_2FA = this->unk_2FB;
+    func_80A75C38(this, globalCtx);
+    if ((this->actor.params == 0) && (this->actor.colChkInfo.health < 11)) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:702]: `<= 10`

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 7]
```diff
@@ -1,15 +1,1401 @@
...
+
+    return displayList;
+}
+
+s32 EnIk_OverrideLimbDraw3(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, Actor* actor) {
+    EnIk* this = (EnIk*)actor;
+
+    if (limbIndex == 0xC) {
+        if (this->actor.params != 0) {
+            *dList = D_06018E78;
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:761]: limbs in decimal

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 8]
```diff
@@ -1,15 +1,1401 @@
...
+    Vec3f spE8;
+    EnIk* this = (EnIk*)actor;
+
+    OPEN_DISPS(globalCtx->state.gfxCtx, "../z_en_ik_inFight.c", 1201);
+
+    if (this->unk_2FB & 1) {
+        func_80032F54(&this->unk_308, limbIndex, 0x1A, 0x1B, 0x1C, dList, -1);
+    }
+    if (limbIndex == 0xC) {
+        gSPMatrix(POLY_XLU_DISP++, Matrix_NewMtx(globalCtx->state.gfxCtx, "../z_en_ik_inFight.c", 1217),
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:812]: in this func too

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 9]
```diff
@@ -1,15 +1,1401 @@
...
+
+void func_80A774F8(EnIk* this, GlobalContext* globalCtx) {
+    if (EnIk_GetNpcAction(globalCtx, 4) == NULL) {
+        Actor_Kill(&this->actor);
+    }
+}
+
+s32 EnIk_OverrideLimbDraw2(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, Actor* actor) {
+    if ((limbIndex == 0xD) || (limbIndex == 0x1A) || (limbIndex == 0x1B)) {
+        if (EnIk_AnimCurrentFrame(actor) >= 30.0f) {
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:1104]: limbs

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 10]
```diff
@@ -1,15 +1,1401 @@
...
+    return 0;
+}
+
+void EnIk_PostLimbDraw2(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3s* rot, Actor* actor) {
+    GraphicsContext* gfxCtx = globalCtx->state.gfxCtx;
+
+    OPEN_DISPS(gfxCtx, "../z_en_ik_inAwake.c", 207);
+
+    switch (limbIndex) {
+        case 0xD:
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:1118]: some more

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c [hunk 11]
```diff
@@ -1,15 +1,1401 @@
...
+
+    sActionFuncs[this->action](this, globalCtx);
+}
+
+s32 EnIk_OverrideLimbDraw1(GlobalContext* globalCtx, s32 limbIndex, Gfx** dList, Vec3f* pos, Vec3s* rot, Actor* actor) {
+    EnIk* this = (EnIk*)actor;
+    f32 animCurrentFrame;
+
+    switch (limbIndex) {
+        case 0x11:
```
Comment (fig02) [inline src/overlays/actors/ovl_En_Ik/z_en_ik.c:1258]: last one ill comment :D

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c (hunk 2 above)
Comment (Sirius902) [reply src/overlays/actors/ovl_En_Ik/z_en_ik.c:404]: Nope, unfortunately not.

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c (hunk 3 above)
Comment (Sirius902) [reply src/overlays/actors/ovl_En_Ik/z_en_ik.c:443]: Yep

File: src/overlays/actors/ovl_En_Ik/z_en_ik.c (hunk 2 above)
Comment (Sirius902) [reply src/overlays/actors/ovl_En_Ik/z_en_ik.c:404]: `unk_2F9` is actually a `u8` but unfortunately storing as / casting to `u8` does a lot of extra floating point operations.

**Review by NEstelami — APPROVED** [review]

**Review by fig02 — APPROVED** [review]

**Pull request merged**